src/LanguageServer/Impl/Implementation/LineFormatter.cs

```diff
@@ -0,0 +1,631 @@
+                        builder.SoftAppendSpace();
+                        builder.Append("\\"); // Hardcoded string so that any following whitespace doesn't make it in.
+                        break;
+
+                    case TokenKind.BackQuote:
+                        builder.Append(token);
+                        break;
+
+                    default:
+                        if (token.Kind == TokenKind.KeywordLambda) {
+                            if (token.IsInsideFunctionArgs && prev?.Kind == TokenKind.Assign) {
+                                builder.Append(token);
+
+                                if (next?.Kind != TokenKind.Colon) {
+                                    builder.SoftAppendSpace();
+                                }
+
+                                break;
+                            }
+                        }
+
+                        if (token.IsKeyword) {
+                            if (prev != null && !prev.IsOpen) {
+                                builder.SoftAppendSpace();
+                            }
+
+                            builder.Append(token);
+
+                            if (next != null && next.Kind != TokenKind.Colon && next.Kind != TokenKind.Semicolon) {
+                                builder.SoftAppendSpace();
+                            }
+
+                            break;
+                        }
+
+                        // No tokens should make it to this case, but try to keep things separated.
+                        builder.SoftAppendSpace();
+                        builder.Append(token);
+                        builder.SoftAppendSpace();
+                        break;
+                }
+            }
+
+            var endCol = _tokenizer.EndOfLineCol(line);
+
+            var afterLast = tokens.Last().Next;
+            if (afterLast != null && afterLast.IsMultilineString) {
+                // If the the next token is a multiline string, then make
+                // sure to include that string's prefix on this line.
+                var afterLastFirst = SplitByNewline(afterLast.ToString()).First();
+                builder.Append(afterLastFirst);
+            }
+
+            var newText = builder.ToString();
+
+            if (newText.Length == 0) {
+                return NoEdits;
+            }
+
+            var edit = new TextEdit {
+                range = new Range {
+                    start = new SourceLocation(line, beginCol),
+                    end = new SourceLocation(line, endCol)
+                },
+                newText = newText
+            };
+
+            return new[] { edit };
+        }
+
+        private class TokenExt {
+            public Token Token { get; set; }
+            public SourceSpan Span { get; set; }
+            public int Line => Span.End.Line;
+            public TokenExt Inside { get; set; }
+            public TokenExt Prev { get; set; }
+            public TokenExt Next { get; set; }
```

Comment:
So `TokenExt` forms a linked list of tokens. How many items it will have on our testing `pythonGrammar.py`?

Reply:
I did a quick modification of the wrapper to count, and the grammar file is 11278 tokens. In the common case, any sort of line formatting would only tokenize up to (and a bit past) the line to be formatted, so usually you wouldn't have this many at once. I'm not sure how much it matters given that the tokenizer essentially reads the whole file into memory. Mikhail said it would even be okay for me to read the whole file into memory (to figure out line lengths) instead of looking at spans to figure out what to replace.

There's definitely a way that I could write this to only look back so far (so that unneeded tokens can get thrown away), but I chose not to do it so that the LineFormatter could potentially be reused and to simplify the initial implementation.

Reply:
Unfortunately, looking back has to start from the beginning. This is somewhat akin to [Kanji backspace](http://discuss.joelonsoftware.com/default.asp?interview.11.334807.4) problem